rfc/account-management.md

```diff
@@ -0,0 +1,135 @@
+
+## Timeline
+
+Spinnaker 1.23 (1.22?):
+- Add interfaces to kork 
+- Start splitting AccountCredentialsRepository by provider (we'd now be able to use multiple accounts with the same name in different providers)
+- AWS, Kubernetes support in Clouddriver.
+
+Future releases:
+- More services/providers as the community sees the need. We expect the change to be non invasive.
+
+We have other providers in mind and are ready to contribute for these as well.
+
+
+## Design
+
+Here's a simple implementation of the interfaces (interfaces should be in kork): [https://github.com/ncknt/clouddriver/compare/master...ncknt:feat/kork-accounts?expand=1](https://github.com/ncknt/clouddriver/compare/master...ncknt:feat/kork-accounts?expand=1)
+
+An `Account` contains the information necessary to access an account. It can be parsed and enriched into a `Credentials`. Spinnaker already has a lot of these: e.g. `com.netflix.spinnaker.clouddriver.security.AccountCredentials`, `com.netflix.spinnaker.clouddriver.artifacts.config.ArtifactCredentials`, `com.netflix.spinnaker.igor.service.BuildService`, etc.
+
+An `AccountParser` translates an `Account` into an instance of `Credentials`. A `CredentialsRepository` contains one type of `Credentials`. Most of the time, we only need access to a single type of credentials (e.g. in clouddriver's providers). When access to different types of credentials is needed (e.g. in clouddriver's artifact), we can use a `CompositeCredentialsRepository`.
+
+Each account repository is wired with an `AccountSource`. The source is queried once when the repository is created but can also be queried again if it is a `ReloadableAccountRepository` which the default implementation is.
+
+If `accounts.dynamic.enabled`, then every `accounts.dynamic.[type of account].reloadFrequencyMs` the source will be queried, parsed, and merged into the repository.
+
+A `CredentialsLifecycleHandler` handles enabling/disabling credentials (e.g. scheduling agents for a new provider account).
+
+Out of the box, nothing should change for users. However, we could now write a plugin that for instance get Kubernetes accounts from a remote endpoint, or polls Jenkins master definitions from a database.
+
+### Clouddriver Providers
+Clouddriver currently uses one repository to store all accounts. Part of the proposal pushes each provider to use a `CredentialsRepository` defined above. It also uses the `AccountCredentialsProvider` abstraction.
+
+For each provider, the migration looks like:
+- Move from `AccountCredentialsProvider` to `CredentialsProvider<T extends AccountCredentials>`
+- `CredentialsProvider` has the same methods except it is using generics
+- Create the default `CredentialsRepository` bean
+- Create the default `CredentialsProvider` bean
+
+`clouddriver-elasticsearch` uses credentials across credentials and can be changed to use a `CompositeCredentialsRepository` that federates `CredentialsRepository` with non migrated provider accounts.
```

Comment:
> uses credentials across credentials

I don't understand this part.

Reply:
Sorry there's a typo! What I meant is that the elasticsearch module seems to need to access credentials across **providers** - or at least it seemed to but I've never used it so it may very well be wrong. I took it as an exercise: what if we need to access multiple types of credentials? Hence the `CompositeCredentialsRepository`.

Reply:
The `clouddriver-elasticsearch` module may not be all that relevant, I don't believe it's widely/if-at-all used outside of Netflix for entity tags. 

There is some lingering experimental indexing code that I suspect can be `rm -rf`'d ... years ago we experimented with maintaining a separate index in elasticsearch but it didn't go anywhere. Removing this code will likely simplify what was thought to be needed here.